src/ast.rs

```diff
@@ -452,6 +208,103 @@
     fn can_be_parsed(pop: &mut Populator<'_>) -> bool;
 }
 
+define_node! {
+    #[derive(Debug)]
+    #[enum_dispatch(Acceptor, Node)]
+    #[enum_dispatch(AcceptorMut, NodeMut)]
+    pub enum Node {
+        Branch,
+        Leaf,
+        List,
```

Comment:
Probably for a future commit/PR:
Can we get rid of `Type` and `Category` now?
They should be mostly redundant (already before but maybe it's easier now).
`ast_type_to_string()` is only used for debugging, it's fine to swap that out with a derived `Debug` impl that simply outputs the type name.

Reply:
Not yet, but it's definitely in the cards. `Category` especially.